### ISSUE TITLE
Rule proposal: One export per file

### ISSUE THREAD

**Issue by devinrhode2** [sindresorhus/eslint-plugin-unicorn#1925]
### Description

I've noticed there are numerous benefits to making files smaller.

1. Reduce merge conflicts. If files are sufficiently small, you usually don't end up shuffling code around within the file. Shuffling code around in a file can easily cause annoying merge conflicts.
2. Smaller tests.
3. Fewer imports per file.

Subjectively, I feel smaller files can improve readability. By breaking code into smaller pieces, I can see smaller chunks of code at a time.

Some people may agree.

As opposed to picking a certain line limit to files (although "1 page" would be reasonable, imo), we can simply limit to 1 export per file.

Once all files have one export, then all import statements are probably going to be one line.
(Unless you are importing stuff from an external npm package)

Since each import is one line, we again reduce the chance of annoying merge conflicts.

Here's an example illustrating merge conflicts around import statements: Imagine a few different git branches:
1.
```tsx
import type { Foo, Bar } from 'types/FooStuff'
```
2.
```tsx
import type { Foo, Baz } from 'types/FooStuff'
```
3.
```tsx
import { bar, Bar, baz, Baz, foo, Foo } from 'types/FooStuff'
```
4.
```tsx
import {
  bar,
  Bar,
  baz,
  Baz,
  foo,
  Foo
} from 'types/FooStuff'
```

I can say from experience, having each imported variable on it's own line makes conflict resolution much easier. <sub>This is a small point, imports already happen automatically, so conflicts on import statements could resolved automatically, once the file doesn't have any other conflicts in the rest of the file.)</sub>

### Fail

```ts
// Foo.ts
export const foo = 'foo'
export const bar = 'bar'
```

```ts
// Foo.ts
const foo = 'foo'
const bar = 'bar'

export { foo, bar }
```

```ts
// Foo.ts
export default const foo = 'foo'
export const bar = 'bar'
```

### Pass

```ts
// Foo.ts
export const foo = 'foo'
```

```ts
export function foo() {
  return 'me';
}
```

```ts
export default const foo = 'foo'
```

**Comment by fisker** [comment]
It's common to put small related functions/constants together in one file.

**Comment by mangs** [comment]
Curious of folks' thoughts on this one.

I'd be a big fan of this with a small tweak: allow only a single `export` statement at the end of a file. I've been doing this manually up until now, but my team and I section our files into sections using comments (e.g. `// External Imports`, `// Internal Imports`, `// Local Functions`, `// Component Definition`, etc.) with the last section being the export. The caveat for us being that we enforce named exports; I don't think this matters in this case. What I'm proposing looks like this (using your example above):

```js
export {
  foo,
  bar,
  // etc.
}
```

The points you made in your original post are some of the reasons why I chose this design in addition to making it easy to find exports in any file: they're always the last thing in any file.

**Comment by fregante** [comment]
Your first suggestion doesn't match your second comment:

1. "One export per file makes code more testable"
2. "I'm only suggesting one `export` keyword"


Having one `export` keyword could help clarify what a file exports in one, compact place. That's kind of a valid reason, even if I would not use it. 

The first suggestion however feels too tight in most cases. 

If it doesn't already exist, perhaps I'd suggest it to `eslint-plugin-import` instead

**Comment by devinrhode2** [comment]
@mangs thats a pretty good idea, much more useful for more people.

I’d say root cause is that there is too much code in a file.

Instead of trying to enforce a single export keyword at the bottom of a file, why not just turn that file into a directory, and each exported name becomes its own file name

**Comment by mangs** [comment]
@devinrhode2 I hear you, and I agree less code per file is good but only to a point. I've found that a good balance is treating a fie as more of a namespace and putting all related functions within it while using named exports (e.g. `utils/environment.mjs` will hold all functions related to dealing with the execution environment (e.g. `export { getUserId }`) instead of having `utils/environment/getUserId.mjs` etc). Default exports with too many files starts to feel like boilerplate management. And tree-shaking eliminates what's not used. Just my thoughts from experience, curious what you think.

**Comment by devinrhode2** [comment]
I really fell hard into this "one exported symbol name per file" idea when I was rebasing/revising this huge stacked PR.

I had a file "TreeNode.ts" which held all types related to.. TreeNodes.

It turned out, damn near half the types in this crazy ui table were related to TreeNodes.

I kept on squashing little tweaks into this file, but I wasn't creating great git history, the history had a too much "knowledge of the future". Anyway, as I went along irresponsibly rebasing, I kept getting rebase merge conflicts on this file. I'm really good/fast at resolving merge conflicts, but I realized I could prevent them outright by splitting my code into more files. By having each type def in it's own file, inside a "TreeNode" folder, I would be able to shuffle/re-order commits more easily, and cause fewer merge conflicts on this one file.

It could also lead to tools like [`Destiny`](https://github.com/benawad/destiny) being supremely awesome
